docs/spec/reactors/consensus/proposer-selection.md

```diff
@@ -2,45 +2,268 @@
+- VP(i) - voting power of validator i
+- P - total voting power of set
+- avg - average of all validator priorities
+- prop - proposer
+
+Simple view at the Selection Algorithm:
+
+```
+    def simpleProposerSelection (vset):
+        for each validator i in vset:
+            A(i) += VP(i)
+        prop = max(A)
+        A(prop) -= P
+```
+
+If the set of the validators has not changed, at the beginning of each new round the sum of their priorities is 0.
+
+Following table shows, for the example above, how the proposer priority is calculated during each round. Only 4 rounds are shown, starting with the 5th the same values are computed.
+Each row shows the imaginary queue and the process place in it. The proposer is the closest to the head of the queue, the rightmost validator. As priorities are updated, the validators move right in the queue. At the end of the round the proposer moves left as its priority is reduced after election. At the end of each round the sum of the priorities is 0.
+
+|Round Priority  | -2| -1| 0   |  1| 2   | 3 | 4 | 5 | Alg step
+|--------------- |---|---|---- |---|---- |---|---|---|--------
+|                |   |   |p1,p2|   |     |   |   |   |Initialized to 0
+|1               |   |   |     | p1|     | p2|   |   |A(i)+=VP(i)
+|                |   | p2|     | p1|     |   |   |   |A(p2)-= P
+|2               |   |   |     |   |p1,p2|   |   |   |A(i)+=VP(i)
+|                | p1|   |     |   |   p2|   |   |   |A(p1)-= P
+|3               |   | p1|     |   |     |   |   | p2|A(i)+=VP(i)
+|                |   | p1|     | p2|     |   |   |   |A(p2)-= P
+|4               |   |   |   p1|   |     |   | p2|   |A(i)+=VP(i)
+|                |   |   |p1,p2|   |     |   |   |   |A(p2)-= P
+
+The actual implementation does not create this queue, it only has to store the current accumulated proposer priority. 
+
+There are cases where the proposer selection is executed multiple times.
+(find a good example here)
+So we modify slightly the algorithm to handle this case. Note that times indicates the number of executions of the simpleProposerSelection() 
+
+    def singleProposerSelection (vset):
+        for each validator i in vset:
+            A(i) += VP(i)
+        prop = max(A)
+        A(prop) -= P
+    
+    def ProposerSelection (vset, times):
+        for in in range(times):
+            singleProposerSelection(vset)
+
+### Validator Set Changes
+At each block height the validator set may change. Some of the changes have implications on the proposer selection.
+
+#### Voting Power Change
+Consider again the earlier example and assume that the voting power of p1 is changed to 4:
+
+
+Validator | p1| p2 
+----------|---| ---
+VP        | 4 | 3
+
+Let's also assume that at the last round R at height H before this change the proposer priorites were as shown in first row (H/R end). As it can be seen, the procedure could continue with round 1 at next height without changes as before. 
```

Comment:
Mentioning the state of the last round at some height is misleading as in the current implementation proposer selections within some height H does not affect selection at height H+1. For a given height H, initial validator set is defined and then for each round priorities are modified. For height H+1, the only thing that matters is the initial validator set of the previous height (H) and potential validator set changes at the end of height H. Between heights H and H+1 we apply the same algorithm for proposer selection that modifies priorities and select the initial proposer of round 0.

Reply:
But the priorities are remembered between blocks, no? We don't reset them AFAIK

Reply:
I must admit I am confused. Even after reading #3044 things are not clear. I just removed everything talking about rounds and heights and only refer to runs of the ProposerSelection(). Also not sure anymore about R1 which says: proposer_p(h,r) = proposer_q(h,r) ...After reading #3044 this is not so obvious to me. So, I left that there as is.

Reply:
R1 is fine, Zarko confirmed and enlightened me on the round incrementPriority and pointed me to the code. I should be able to figure it out. For the spec, I removed the mentions of the rounds.